src/install-opentofu.sh

```diff
@@ -0,0 +1,1206 @@
+      if is_root; then
+        log_debug "We are already root, no user change needed."
+        "$@"
+      elif command_exists "sudo"; then
+        log_debug "Running command using sudo."
+        sudo "$@"
+      elif command_exists "su"; then
+        log_debug "Running command using su."
+        su root "$@"
+      else
+        log_error "Neither su nor sudo is installed, cannot obtain root privileges."
+        return $TOFU_INSTALL_RETURN_CODE_COMMAND_NOT_FOUND
+      fi
+      return $?
+      ;;
+    none)
+      log_debug "Using manual root method 'none'."
+      "$@"
+      return $?
+      ;;
+    sudo)
+      log_debug "Using manual root method 'sudo'."
+      sudo "$@"
+      return $?
+      ;;
+    su)
+      log_debug "Using manual root method 'su'."
+      su root "$@"
+      return $?
+      ;;
+    *)
+      log_error "Bug: invalid root method value: $1"
+      return $TOFU_INSTALL_EXIT_CODE_INVALID_ARGUMENT
+  esac
+}
+
+# This function attempts to execute a function as the current user and switches to root if it fails.
+maybe_root() {
+  if ! "$@" >/dev/null 2>&1; then
+    if ! as_root "$@"; then
+      return $TOFU_INSTALL_EXIT_CODE_INSTALL_FAILED
+    fi
+  fi
+  return $TOFU_INSTALL_EXIT_CODE_OK
+}
+
+# This function verifies if one of the supported download tools is installed and returns with
+# $TOFU_INSTALL_EXIT_CODE_DOWNLOAD_TOOL_MISSING if that is not th ecase.
+download_tool_exists() {
+    log_debug "Determining if a supported download tool is installed..."
+    if command_exists "wget"; then
+      log_debug "wget is installed."
+      return $TOFU_INSTALL_EXIT_CODE_OK
+    elif command_exists "curl"; then
+      log_debug "curl is installed."
+      return $TOFU_INSTALL_EXIT_CODE_OK
+    else
+      log_debug "No supported download tool is installed."
+      return $TOFU_INSTALL_EXIT_CODE_DOWNLOAD_TOOL_MISSING
+    fi
+}
+
+# This function downloads the URL specified in $1 into the file specified in $2.
+# It returns $TOFU_INSTALL_EXIT_CODE_DOWNLOAD_TOOL_MISSING if no supported download tool is installed, or $TOFU_INSTALL_RETURN_CODE_DOWNLOAD_FAILED
+# if the download failed.
+download_file() {
+  if [ -z "$1" ]; then
+    log_error "Bug: no URL supplied to download_file()"
+    return $TOFU_INSTALL_EXIT_CODE_INVALID_ARGUMENT
+  fi
+  if [ -z "$2" ]; then
+    log_error "Bug: no destination file supplied to download_file()"
+    return $TOFU_INSTALL_EXIT_CODE_INVALID_ARGUMENT
+  fi
+  log_debug "Downloading URL ${1} to ${2}..."
+  IS_GITHUB=0
+  if [ -n "${GITHUB_TOKEN}" ]; then
+    if [ "$(echo "$1" | grep -c "api.github.com")" -ne 0 ]; then
+      IS_GITHUB=1
+    fi
+  fi
+  if command_exists "wget"; then
+    log_debug "Downloading using wget..."
+    if [ "${IS_GITHUB}" -eq 1 ]; then
+      if ! wget -q --header="Authorization: token ${GITHUB_TOKEN}" -O "$2" "$1"; then
+        log_debug "Download failed."
+        return $TOFU_INSTALL_RETURN_CODE_DOWNLOAD_FAILED
+      fi
+    else
```

Comment:
Might be nice to have a log in here mentioning that it's using the github token because it was detected.

Otherwise some people may get confused if it fails